pkg/ovs/openflow/ofctrl_bridge.go

```diff
@@ -262,6 +263,89 @@ func (b *OFBridge) IsConnected() bool {
 	return b.ofSwitch.IsReady()
 }
 
+func (b *OFBridge) AddFlowsInBundle(addflows []Flow, modFlows []Flow, delFlows []Flow) error {
+	// If none Openflow entries are requested to be added or modified or deleted on OVS, return.
+	if addflows == nil && modFlows == nil && delFlows == nil {
+		klog.V(2).Infoln("None Openflow entries are needed to sync to OVS, return.")
+		return nil
+	}
+	// Create a new transaction.
+	tx := b.ofSwitch.NewTransaction(ofctrl.Atomic)
+	// Open a bundle on the OFSwitch.
+	if err := tx.Begin(); err != nil {
+		return err
+	}
+
+	syncFlows := func(flows []Flow, operation int) error {
+		for _, flow := range flows {
+			ofFlow := flow.(*ofFlow)
+			ofFlow.Flow.NextElem = ofFlow.lastAction
+			// "AddFlow" operation is async, the function returns error when constructing and sending BundleAdd message.
+			// It doesn't represent all Openflow entries are added into the bundle even if the functionreturns nil.
+			// The number of the Openflow entries successfully added into the bundle is returned from function "Complete".
+			flowMod, err := ofFlow.Flow.GenerateFlowModMessage(operation)
+			if err != nil {
+				return err
+			}
+			if err := tx.AddFlow(flowMod); err != nil {
+				// Close the bundle and abort it if there is error when adding the FlowMod message.
+				_, err := tx.Complete()
+				if err == nil {
+					tx.Abort()
+				}
+				return err
+			}
+		}
+		return nil
+	}
+
+	// Install new Openflow entries with the opened bundle.
+	if err := syncFlows(addflows, openflow13.FC_ADD); err != nil {
+		return err
+	}
+	// Modify existing Openflow entries with the opened bundle.
+	if err := syncFlows(modFlows, openflow13.FC_MODIFY_STRICT); err != nil {
+		return err
+	}
+	// Delete Openflow entries with the opened bundle.
+	if err := syncFlows(delFlows, openflow13.FC_DELETE_STRICT); err != nil {
+		return err
+	}
+
+	// Close the bundle before commit it to the OFSwitch.
+	count, err := tx.Complete()
+	if err != nil {
+		return err
+	} else if count != len(addflows)+len(modFlows)+len(delFlows) {
```

Comment:
what's the behavior for flows which are already installed and for which we call "add" again?

Reply:
I think the already existed flows should not "added" again, and it should be decided by the caller. For a bundle, it succeeds only if all the messages(add/modify/delete) are retained by OVS, otherwise, it returns error means none of the flow entries is retained on the OVS.

So, the client could know which flows are already installed, and which are not according the bundle result. And there should be chance that partial successfully.